### PR TITLE
CRIU delay Thread.interrupt() in single thread mode

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Added `J9_SINGLE_THREAD_MODE_OP_INTERRUPT`;
Saved Thread instances to `delayedLockingOperationsRoot` queue;
`CRIUHelpers` refactoring ;
Added tests.

closes https://github.com/eclipse-openj9/openj9/issues/15538

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: runtime/oti/VMHelpers.hpp [hunk 1]
```diff
@@ -2012,6 +2013,20 @@ class VM_VMHelpers
...
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	/*
+	 * Queries whether the JVM is running in single thread mode
+	 *
+	 * @param[in] vm Java VM token
+	 * @return True if JVM is in single thread mode, false otherwise
+	 */
+	static VMINLINE bool
+	isJVMInSingleThreadMode(J9JavaVM *vm)
```
Comment (tajila) [inline runtime/oti/VMHelpers.hpp:2024]: why is this moved?

File: runtime/vm/CRIUHelpers.cpp [hunk 2]
```diff
@@ -476,4 +495,44 @@ runDelayedLockRelatedOperations(J9VMThread *currentThread)
 	return rc;
 }
 
+BOOLEAN
+delayedLockingOperation(J9VMThread *currentThread, j9object_t instance, UDATA operation)
```
Comment (tajila) [inline runtime/vm/CRIUHelpers.cpp:499]: we can leave this where it is and add a c wrapper for it. Example is in ValueTypeHelpers.cpp

File: runtime/vm/CRIUHelpers.cpp [hunk 3]
```diff
@@ -413,8 +413,26 @@ runDelayedLockRelatedOperations(J9VMThread *currentThread)
 	Assert_VM_true(vm->checkpointState.checkpointThread == currentThread);
 
 	while (NULL != delayedLockingOperation) {
-		omrthread_monitor_t monitorPtr = NULL;
 		j9object_t instance = J9_JNI_UNWRAP_REFERENCE(delayedLockingOperation->globalObjectRef);
+		omrthread_monitor_t monitorPtr = NULL;
+		if (J9_SINGLE_THREAD_MODE_OP_INTERRUPT == delayedLockingOperation->operation) {
+			J9VMThread *targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, instance);
```
Comment (tajila) [inline runtime/vm/CRIUHelpers.cpp:419]: can we add a common helper for this, so the code isnt dplicated?

File: runtime/oti/VMHelpers.hpp (hunk 1 above)
Comment (JasonFengJ9) [reply runtime/oti/VMHelpers.hpp]: Restored, was thinking to take out `CRIUHelpers.hpp`, apparently isn't right direction.

File: runtime/vm/CRIUHelpers.cpp (hunk 3 above)
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp]: Added `VM_VMHelpers::threadInterruptImpl`.

File: runtime/vm/CRIUHelpers.cpp (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp:488]: Replaced it w/ `VM_CRIUHelpers::delayedLockingOperation()`.

**Review by tajila — COMMENTED** [review]
In `monitorWaitImpl` we should also detect if we are interupted in single thread mode due to spurious interupts and wait again.

**Comment by JasonFengJ9** [comment]
Updated `monitorWaitImpl` & `jvmtiRawMonitorWait` to handle spurious interruptions.
@tajila this is ready for another look.

File: runtime/vm/threadhelp.cpp [hunk 4]
```diff
@@ -119,32 +118,73 @@ monitorWaitImpl(J9VMThread *vmThread, j9object_t object, I_64 millis, I_32 nanos
...
+	) {
+		if (waitTimed) {
+			PORT_ACCESS_FROM_JAVAVM(javaVM);
+			const I_32 oneMillion = 1000000;
+			I_64 waitedTimeNanos = j9time_nano_time() - beforeWait;
+			I_64 waitedTimeMillis = waitedTimeNanos / oneMillion;
+			if (waitedTimeMillis < millis) {
+				millis -= waitedTimeMillis;
+				waitedTimeNanos = (waitedTimeNanos % oneMillion);
+				if (waitedTimeNanos >= nanos) {
```
Comment (tajila) [inline runtime/vm/threadhelp.cpp:150]: what if user only specifies `wait(0, something)` ? doesnt the nanoSecond check need to be done independantly of the mili second?

File: runtime/jcl/common/thread.cpp [hunk 5]
```diff
@@ -277,23 +277,19 @@ Java_java_lang_Thread_interruptImpl(JNIEnv *env, jobject rcv)
 	J9VMThread *currentThread = (J9VMThread*)env;
 	J9JavaVM *vm = currentThread->javaVM;
 	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	j9object_t receiverObject = J9_JNI_UNWRAP_REFERENCE(rcv);
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/jcl/common/thread.cpp]: Yeah, fixed.

File: runtime/vm/threadhelp.cpp (hunk 4 above)
Comment (JasonFengJ9) [reply runtime/vm/threadhelp.cpp:150]: If less than `1ms` left for monitor waiting (either `wait(0, something)` or just the `millis` specified passed), I think there is a chance that this thread will be timed out and active while single thread mode is still on, hence waiting another `10ms` to accommodate such cases, the number was picked arbitrarily, I am open to a different timeout or nanos specified instead.

**Comment by tajila** [comment]
Jenkins test sanity xlinuxcriu jdk17

**Comment by tajila** [comment]
failures are known failures
